examples/call-service-panel-example/src/CallServicePanel.tsx

```diff
@@ -0,0 +1,112 @@
+import { PanelExtensionContext, RenderState } from "@foxglove/studio";
+import { useCallback, useEffect, useLayoutEffect, useState } from "react";
+import ReactDOM from "react-dom";
+import ReactJson from "react-json-view";
+
+type State = {
+  serviceName: string;
+  request: string;
+  response?: unknown;
+  error?: Error | undefined;
+  colorScheme?: RenderState["colorScheme"];
+};
+
+function CallServicePanel({ context }: { context: PanelExtensionContext }): JSX.Element {
+  const [renderDone, setRenderDone] = useState<(() => void) | undefined>();
+  const [state, setState] = useState<State>({ serviceName: "", request: "{}" });
+
+  useLayoutEffect(() => {
+    context.onRender = (renderState: RenderState, done) => {
+      setState((oldState) => ({ ...oldState, colorScheme: renderState.colorScheme }));
+      setRenderDone(() => done);
+    };
+  }, [context]);
+
+  context.watch("colorScheme");
+
+  useEffect(() => {
+    renderDone?.();
+  }, [renderDone]);
+
+  const callService = useCallback(
+    async (serviceName: string, request: string) => {
+      if (!context.callService) {
+        return;
+      }
```

Comment:
Is there a way to re-render the panel when `context.callService` changes? Right now I have to remove and re-add the panel after connecting to a foxglove websocket connection. Maybe the problem is, that the fg-ws player does not know before receiving the first server message if it can declare the callService capability?

Reply:
In Studio we re-mount all the panels (which will re-initialize) extension panels when the user selects a new connection. However, I see that we don't actually set a `callService` function until the capability is available: https://github.com/foxglove/studio/blob/main/packages/studio-base/src/components/PanelExtensionAdapter/PanelExtensionAdapter.tsx#L440

What you can do to re-render the panel (if using react) is to track the callService value. If you detect that the value has changed (i.e. went from undefined -> defined). In your code you could maybe update the state based on whether it has changed or not to cause a re-render.

If what I described is not clear I can pair to show how this could work with this example panel to achieve the UX you want.

Reply:
I tried that. For debugging, I added a 
```
  useEffect(() => {
    console.log(`context.callService has changed: ${context.callService}`);
  }, [context.callService]);
```
and what is logged confirms that `context.callService` does not update:

```
Opening connection to ws://localhost:8765
Init panel d2a28246-bf7a-4cf6-bdce-f38c0b54bcac
context.callService has changed: undefined
```

My hunch is, that the top-level component (which renders the panel) does not detect itself that `context.callService` has changed and hence it does not re-render panels.

Reply:
:point_up: I could just confirm that. However, the actual reason is, that the change in player [capabilities](https://github.com/foxglove/studio/blob/ca36c94a3c07db0467d36085415c07520a50550f/packages/studio-base/src/components/PanelExtensionAdapter/PanelExtensionAdapter.tsx#L479) is not detected and hence `context.callService` is not updated.

It works correctly, if I change the [capability](https://github.com/foxglove/studio/blob/ca36c94a3c07db0467d36085415c07520a50550f/packages/studio-base/src/components/PanelExtensionAdapter/PanelExtensionAdapter.tsx#L479) dependency from
```ts
  }, [
    capabilities,
    ...
  ]
```

to
```ts
  }, [
    // eslint-disable-next-line react-hooks/exhaustive-deps
    JSON.stringify(capabilities),
    ...
  ]
```

Will create an issue on foxglove/studio for that